File: homework4_1/src/main/java/by/itacademy/homework4_1/MainActivity.java
package by.itacademy.homework4_1;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.SearchView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MainActivity extends AppCompatActivity implements IObserver {
    private ItemAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
        RecyclerView recycler = findViewById(R.id.recycler);
        setRecycler(recycler);
        Publisher.getInstance().addSubscriber(this);

        findViewById(R.id.addItem).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(MainActivity.this, CreateItemActivity.class));
            }
        });
        SearchView searchView = findViewById(R.id.searchContact);
        searchViewClickListener(searchView);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Publisher.getInstance().removeSubscriber(this);
    }

    @Override
    public void notifyDataChange(int position, int operation) {
        switch (operation) {
            case OperationType.REMOVE: {
                adapter.notifyItemRemoved(position);
                break;
            }
            case OperationType.CHANGE: {
                adapter.notifyItemChanged(position);
                break;
            }
        }
    }

    private void setRecycler(RecyclerView recycler) {
        adapter = new ItemAdapter(Publisher.getInstance().getContactList(), new ItemAdapter.ListItemActionListener() {
            @Override
            public void onItemClicked(int number) {
                Intent intent = new Intent(MainActivity.this, ChangeItemActivity.class);
                intent.putExtra("position", number);
                startActivity(intent);
            }
        });
        recycler.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        recycler.setAdapter(adapter);
    }

    private void searchViewClickListener(SearchView searchView) {
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                if (newText.equals("")) {
                    adapter.setContacts(Publisher.getInstance().getContactList());
                } else {
                    adapter.filter(newText);
                }
                return true;
            }
        });
    }

}

File: homework5_2/src/main/java/by/itacademy/homework5_2/MainActivity.kt
package by.itacademy.homework5_2

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.SearchView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import by.itacademy.homework5_2.data.Contact
import by.itacademy.homework5_2.data.DBOperationsImpl
import by.itacademy.homework5_2.databinding.ActivityMainBinding
import by.itacademy.homework5_2.multithread.UsersListListener
import by.itacademy.homework5_2.multithread.MultiThreadFactory
import by.itacademy.homework5_2.multithread.MultiThreadOperations
import by.itacademy.homework5_2.multithread.RXJAVA_HELPER

class MainActivity : AppCompatActivity(), ListItemActionListener, UsersListListener {
    private lateinit var binding: ActivityMainBinding
    private lateinit var itemAdapter: ItemAdapter
    private val threadHelper: MultiThreadOperations by lazy { MultiThreadFactory().createHelper(this, RXJAVA_HELPER) }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)
        search()
        buttonClick(binding.addItem)
        setRecycler(this)
    }

    override fun onResume() {
        super.onResume()
        threadHelper.getUsersFromDB(this)
    }

    private fun buttonClick(button: Button) {
        button.setOnClickListener {
            startActivity(Intent(this, CreateItemActivity::class.java))
        }
    }

    private fun setRecycler(activity: MainActivity) {
        val recycler = binding.recycler
        itemAdapter = ItemAdapter(activity)
        recycler.apply {
            layoutManager = LinearLayoutManager(activity)
            adapter = itemAdapter
        }
    }

    private fun search() {
        binding.searchContact.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                return true
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                val newList = DBOperationsImpl(this@MainActivity).getUsersFromDB() //как избавиться от этого костыля??
                        .filter { contact -> contact.name.toLowerCase().contains(newText.toString().toLowerCase()) }
                itemAdapter.updateItem(newList)
                return true
            }
        })
    }

    override fun onItemClicked(number: Int) {
        intent = Intent(this, ChangeItemActivity::class.java)
        intent.putExtra("position", number)
        startActivity(intent)
    }

    override fun getUsersList(list: List<Contact>) {
        itemAdapter.updateItem(list)
    }
}

